[CNSMR-1991] How to deal with Localizations (#174)
Babylonpartners/ios-playbook: Cookbook/Technical-Documents/Localizations.md

    
- `Base language value`: The actual string value corresponding to the first language in the list. Placeholders are supported with the `%@` where dynamic values are expected. 

Note: the translations will be done by a third party that has access to lokalise and it should be done and notified by your squad.

### Edit a key/value
Search and select the key you'd like to update and change any language you need. The languages you may not have the values will be marked as `Not-verified` and the translators will take care.

## How to add a new Language/Locale

### 1. Xcode: create Files if needed
- Select the Babylon Project / Info / Localizations then click on the `+` button and choose the Language/Locale you want to add.
![Create file, if needed](Assets/Localizations-createFile.png)

- Select the `.strings` file in the Project Navigator, review it in the File Inspector / Localization and select the language/locale you want.
- In `Fastlane/Lanes/lokalise`, add the language/locale code (`en_US` for example) in the `langs` parameter for the desired target. `langs: 'es_US,en_US'` (**comma separated without space**). 

> ### 2. Lokalise.co

Reviewer: ```suggestion
### 2. Lokalise
```

> - If the project still doesn't exist, click on the `+` plus button beside other flags. Find the desired language on the list, and then add it.

Reviewer: ```suggestion
- If the project doesn't exist, click on the `+` plus button beside other flags. Find the desired language on the list, then add it.
```

Note: someone else from your squad should have done this before.

#### 2.1 Populate the files - pull from lokalise
- Check [Lokalise pull guide](https://github.com/Babylonpartners/ios-playbook/blob/master/Cookbook/Technical-Documents/Lokalise.md).

#### 2.2 Commit just your changes
Stage just your changes (additions and editions) on git. Don't stage any additions nor change that you don't recognize. It's really nice to be proactive but not in this case. It might cause unexpected/premature changes. Yes, discard others's changes.

## Target specific localizable - Lokalise project and `strings` files

As the name suggests, it's a `.strings` file that contains texts intended to be present **only** in a specific app target.
The values in this file will be priority over the main file value.

> <br><br>

Reviewer: `<br>` tags should be avoided in markdown ([ref](https://github.com/google/styleguide/blob/gh-pages/docguide/style.md#trailing-whitespace))

```suggestion

```

> E.g: <br>

Reviewer: ```suggestion
E.g:
```

> In `Babylon/Supporting Files/Localization/Localizable.strings`:<br>

Reviewer: ```suggestion
In `Babylon/Supporting Files/Localization/Localizable.strings`:
```

> ```"consultation_gp_details_name" = "GP Name:";```<br>

Reviewer: ```suggestion
`"consultation_gp_details_name" = "GP Name:";`
```

> <br>

Reviewer: ```suggestion

```

> And in `Babylon/Brand/Telus/TargetSpecificLocalizable.strings`:<br>

Reviewer: ```suggestion
And in `Babylon/Brand/Telus/TargetSpecificLocalizable.strings`:
```

> ```"consultation_gp_details_name" = "Doctor Name:";```<br>

Reviewer: ```suggestion
`"consultation_gp_details_name" = "Doctor Name:";`
```

> <br>

Reviewer: ```suggestion

```

When we run the target **Telus**, the value will be taken from the target specific file.
`Doctor Name:`

Whereas when we run **Babylon** the value will be the main one. 
`GP Name:`

This prioritization happens in the static method [`LocalizationCore.localizedName`](https://github.com/Babylonpartners/babylon-ios/blob/develop/SDK/BabylonCore/BabylonCore/Utilities/LocalizationCore.swift#L14) that basically tries first to get the value for such key from the target-specific file and if the result is still the key itself, then it falls back to the main value. 

```swift
public static func localizedName(key: String, bundle: Bundle = Bundle.main) -> String {
    let targetLocalizationString = NSLocalizedString(key, tableName: "TargetSpecificLocalizable", comment: "")
    let targetHasKey = targetLocalizationString != key
    return targetHasKey ? targetLocalizationString : NSLocalizedString(key, bundle: bundle, comment: "")
}
```

> Such approach allows other targets to have their custom keys' values when, for some reason, the user uses the device in British English.

Reviewer: ```suggestion
Such approach allows other targets to have their custom key values when, for some reason, the user uses the device in British English.
```

> Not only that, but also avoids us to have useless and risky repetitions like:<br>

Reviewer: ```suggestion
Not only that, but also avoids us to have useless and risky repetitions like:
```

> ```"consultation_gp_details_name_babylon" = "GP Name:";```<br>

Reviewer: Can't add suggestion because of the syntax 🔥 This should be:

\```
"consultation_gp_details_name_babylon" = "GP Name:";

(Note the code block starts before the string and not closed ,`<br>` removed)

> ```"consultation_gp_details_name_telus" = "Doctor Name:";```<br>

Reviewer: Can't suggest again 🔥 

"consultation_gp_details_name_telus" = "Doctor Name:";
\```

(Note the code block not opened, that's already done in the previous line, but closed in the following line. `<br>` removed)

in the same file released for all apps. 

### How to create a key/value that goes to this file? (lokalise key configuration + fastlane output file location)
Considering you are working in an app (different than the main Babylon UK). A project in Lokalise is needed. Then when you need to have a custom value for a certain key. 

> #### In www.lokalise.co:

Reviewer: ```suggestion
#### In Lokalise
```

3. Click on `Advanced` and in `Assigned to file` select in the dropdown `TargetSpecificLocalizable.strings`. If it's the first time, `New file...` then add the before mentioned `TargetSpecificLocalizable.strings`.

#### In `Babylon/fastlane/Lanes/lokalise`:
1. Create a lane that will pull the string values into the custom file:
Consider that the app name here is `FooBar` and that you have the real `projectId` from Lokalise:
```ruby
# lokalise_pull FooBar
lane :lokalise_pull_foobar do
	lokalise_pull(
		projectId: "0000000000000000000000.11111111", 
		destinationFolder: "../Babylon/Brand/FooBar/",	
		noLanguageFolders: "1", 
		langs: "en, fr"
	)
end
```
2. The detail to highlight here is the `destinationFolder: "../Babylon/Brand/FooBar/"` combined with the given `fileName` in each key entry, it'll compose the full path of the custom file. 

3. The last detail to take care are the `langs` that should reflect the `APP_LANGUAGES` array in the app target's Build Settings / User defined. 

4. Then finally run the brand new command: 
```sh
bundle exec fastlane lokalise_pull_foobar
```
The new target specific file should appear in your git diff. ✨

> ## Lokalise: add comments and screenshots when possible.

Reviewer: Should we mention that comments shouldn't be committed in the `.strings` file?

Author: hmm, that we could solve in the `pull` command to lokalise. 🙈 
We need a clean in the strings to remove them, and add such param to avoid'em.

> Comments and screenshots are important complementary **contexts** for the translators. For that, in the key/value edition screen (on lokalise.co) go to `Advanced` tab, and add a comment on the `Context` field. To upload screenshots, click on the "photo" icon in the keys list and upload them. 

Reviewer: ```suggestion
Comments and screenshots are important complementary **contexts** for the translators. For that, in the key/value edition screen (on Lokalise) go to `Advanced` tab, and add a comment on the `Context` field. To upload screenshots, click on the "photo" icon in the keys list and upload them. 
```

> Imagine that when the translators started the Arabic work, they saw the call-to-action `Book` for Book an appointment screen. Then without any comment nor screenshot, that has been translated **litterally** as book 📚📖 (the object that has pages that we read). As an Engineer you should understand that context is crucial. A quick comment and screenshot made in 5 minutes, might save hours of back and forward to fix a misunderstanding of meaning throughout engineering, management, translators and in this case, AppStore approval proccess. 

Reviewer: ```suggestion
Imagine that when the translators started the Arabic work, they saw the call-to-action `Book` for Book an appointment screen. Then without any comment nor screenshot, that has been translated **literally** as book 📚📖 (the object that has pages that we read). As an Engineer you should understand that context is crucial. A quick comment and screenshot made in 5 minutes, might save hours of back and forward to fix a misunderstanding of meaning throughout engineering, management, translators and in this case, AppStore approval proccess. 
```

  - Canadian English
  - Canadian French (soon)

Babylon US:
  - American English
  - American Spanish

## Build Settings
### Understanding `app_languages` variable
Given that, we manage our languages in an array in Build Settings in each target.

| Babylon UK | Babylon US | Telus |
|--------|--------|--------|
| `en.lproj/*` | `en-US.lproj/*` `es-US.lproj/*` | `en.lproj/*` |

**I.E we control the languages for each target editing the `APP_LANGUAGES` under _User-Defined_**

### _Excluded Source File Names_ and _Included Source File Names_ under `Build option`
In the _Excluded Source File Names_ we exclude any language (`.lproj/`) previously set:
`*.lproj/*` 

In the _Included Source File Names_ we include the languages (`.lproj/`) set in the above mentioned `app_languages` array:
`${APP_LANGUAGES}` in Babylon US for example it is en-US.lproj/* es-US.lproj/*

## How to add a new or changing an existing key/value
### Add a new key/value
In Lokalise, select the corresponding project. 
Click on the `Add key ⌘K` button
- `key` - Give a name for the key following the bread-crumb style. For example: `add_family_member_email_placeholder`, `biometric_touchid_primer_description`.
    

> - `Base language value`: The actual string value corresponding to the first language in the list. Placeholders are supported with the `@%` where dynamic values are expected. 

Reviewer: 👀 
```suggestion
- `Base language value`: The actual string value corresponding to the first language in the list. Placeholders are supported with the `%@` where dynamic values are expected. 
```

# How to deal with Localizations

In project we can add a new Localization. But that is not in target level. Babylon needs to have a different sets of languages/locales per product.

Babylon:
  - British English
  - Arabic (soon), etc 

> NHS111 and Bupa:

Reviewer 2: ```suggestion
NHS111:
```

Author: I made a few more here https://github.com/babylonhealth/ios-playbook/pull/174/commits/3f3015a3d9c9511612655716b067412a60541402 thanks!

Author: I made a few more here https://github.com/babylonhealth/ios-playbook/pull/174/commits/3f3015a3d9c9511612655716b067412a60541402 thanks 👍

`${APP_LANGUAGES}` in Babylon US for example it is en-US.lproj/* es-US.lproj/*

## How to add a new or changing an existing key/value
### Add a new key/value
In Lokalise, select the corresponding project. 
Click on the `Add key ⌘K` button
- `key` - Give a name for the key following the bread-crumb style. For example: `add_family_member_email_placeholder`, `biometric_touchid_primer_description`.
    
- `Base language value`: The actual string value corresponding to the first language in the list. Placeholders are supported with the `%@` where dynamic values are expected. 

### Edit a key/value
Search and select the key you'd like to update and change any language you need. The languages you may not have the values will be marked as `Not-verified` and the translators will take care.

## How to add a new Locale
- In Lokalise project, click on the `+` plus button beside other flags. Find the desired language on the list, and then add it.
- In the iOS project, in the file `lokalise` (under the fastlane folder), add the language code (`en_US` for example) in the `langs` parameter for the desired target. `langs: 'es_US,en_US'` (comma separated without space). 

> 

Reviewer 2: I think there should be a third point here, similar-ish to the below:

```suggestion
- In Xcode - go to the Babylon Project, under localizations select the specific language you want to add to the project
<img width="674" alt="Screenshot 2020-01-07 at 14 40 16" src="https://user-images.githubusercontent.com/24831493/71903320-b794f900-315b-11ea-82c9-d6c0b947a7c7.png">

```

Author: we made this 👍

Reviewer 2: Resolving this - @daniloaliberti fixed it in his latest commits

## Pull from lokalise

> Check [Lokalise pull guide](https://github.com/Babylonpartners/ios-playbook/blob/master/Cookbook/Technical-Documents/Lokalise.md)

Reviewer 3: ```suggestion
Check [Lokalise pull guide](https://github.com/Babylonpartners/ios-playbook/blob/master/Cookbook/Technical-Documents/Lokalise.md).
```

## Commit just your changes

> Stage just your changes (additions and editions) on git. Don't stage any additions nor change that you don't recognize. It's really nice to be proactive but not in this case. It might cause unspected/premature changes. Yes, discard others's changes.

Reviewer 3: [Nit] Could these 2 sections be merged in to one called "Syncing Lokalise changes" or something similar? Then make 'Pull from Lokalise' and 'Commit you changes' as numbered steps?

Author: https://github.com/babylonhealth/ios-playbook/pull/174/commits/a8b681b2a8eceb5fb26e6876ae1a2ff58657dc08 👍 it's a bit different. Let me know what do you think?

## Target specific localizable - Lokalise project and `strings` files

As the name suggests, it's a `.strings` file that contains texts intended to be present **only** in a specific app target.
The values in this file will be priority over the main file value.
<br><br>
E.g: <br>
In `Babylon/Supporting Files/Localization/Localizable.strings`:<br>
```"consultation_gp_details_name" = "GP Name:";```<br>
<br>
And in `Babylon/Brand/Telus/TargetSpecificLocalizable.strings`:<br>
```"consultation_gp_details_name" = "Doctor Name:";```<br>
<br>
When we run the target **Telus**, the value will be taken from the target specific file.
`Doctor Name:`

Whereas when we run **Babylon** the value will be the main one. 
`GP Name:`

> This prioritization happens in the static method [`LocalizationCore.localizedName`](https://github.com/Babylonpartners/babylon-ios/blob/develop/SDK/BabylonCore/BabylonCore/Utilities/LocalizationCore.swift#L14) that basically tries first to get the value for such key from the Specific file and if the result is still the key itself, then it fallsback to the main value. 

Reviewer 3: ```suggestion
This prioritization happens in the static method [`LocalizationCore.localizedName`](https://github.com/Babylonpartners/babylon-ios/blob/develop/SDK/BabylonCore/BabylonCore/Utilities/LocalizationCore.swift#L14) that basically tries first to get the value for such key from the target-specific file and if the result is still the key itself, then it falls back to the main value. 
```

```swift
public static func localizedName(key: String, bundle: Bundle = Bundle.main) -> String {
    let targetLocalizationString = NSLocalizedString(key, tableName: "TargetSpecificLocalizable", comment: "")
    let targetHasKey = targetLocalizationString != key
    return targetHasKey ? targetLocalizationString : NSLocalizedString(key, bundle: bundle, comment: "")
}
```

> Such approach allows other targets to have treir custom keys' values when, for some reason, the user uses the device in British English.

Reviewer 3: ```suggestion
Such approach allows other targets to have their custom keys' values when, for some reason, the user uses the device in British English.
```

#### In `Babylon/fastlane/Lanes/lokalise`:
1. Create a lane that will pull the string values into the custom file:
Consider that the app name here is `FooBar` and that you have the real `projectId` from Lokalise:
```ruby
# lokalise_pull FooBar
lane :lokalise_pull_foobar do
	lokalise_pull(
		projectId: "0000000000000000000000.11111111", 
		destinationFolder: "../Babylon/Brand/FooBar/",	
		noLanguageFolders: "1", 
		langs: "en, fr"
	)
end
```
2. The detail to highlight here is the `destinationFolder: "../Babylon/Brand/FooBar/"` combined with the given `fileName` in each key entry, it'll compose the full path of the custom file. 

3. The last detail to take care are the `langs` that should reflect the `APP_LANGUAGES` array in the app target's Build Settings / User defined. 

4. Then finally run the brand new command: 
```sh
bundle exec fastlane lokalise_pull_foobar
```
The new target specific file should appear in your git diff. ✨

## Lokalise: add comments and screenshots when possible.

> Comments and screenshots are important complementary **contexts** for the translators.

Reviewer 3: Good idea ✨ Might it be worth explicitly mentioning the 'Context' field in the 'Advanced' section in the Lokalise entry as a place to write text / paste image links?

Author: https://github.com/babylonhealth/ios-playbook/pull/174/commits/4b7e00bfbe190f2d1eb81145dc4653e8605bd865 thanks

# How to deal with Localizations [draft]

In project we can add a new Localization. But that is not in target level. Babylon needs to have a different sets of languages/locales per product.

Babylon:
  - British English
  - Arabic (soon), etc 

NHS111 and Bupa:
  - British English

Telus:
  - Canadian English
  - Canadian French (soon)

Babylon US:
  - American English
  - American Spanish

## Build Settings
### Understanding `app_languages` variable
Given that, we manage our languages in an array in Build Settings in each target.

| Babylon UK | Babylon US | Telus |
|--------|--------|--------|
| `en.lproj/*` | `en-US.lproj/*` `es-US.lproj/*` | `en.lproj/*` |

> **I.E we control the languages for each target editing the `APP_LANGUAGES` under _User-Defined_**

Reviewer 4: Please add screenshots on this area, for more junior devs to understand what this means.

Author: My plan is to add screenshots in all sections. It's an outline (a draft, with a few content).

Author: ![Screen Shot 2019-07-04 at 15 56 21](https://user-images.githubusercontent.com/17405823/60727612-3b182800-9f36-11e9-9490-1d62e3594d62.png)

### _Excluded Source File Names_ and _Included Source File Names_ under `Build option`
In the _Excluded Source File Names_ we exclude any language (`.lproj/`) previously set:
`*.lproj/*` 

In the _Included Source File Names_ we include the languages (`.lproj/`) set in the above mentioned `app_languages` array:

> `${APP_LANGUAGES}` in Babylon US for example it is en-US.lproj/* es-US.lproj/*

Reviewer 4: I don't fully understand what this is. 😞

Author: The full content will explain in details. With Xcode screenshots.

Author: <img width="974" alt="Screen Shot 2019-07-05 at 14 39 53" src="https://user-images.githubusercontent.com/17405823/60726183-df986b00-9f32-11e9-86da-bd96e2fd52f3.png">

With the target Babylon US selected, on Build settings / Build options / Include Source File Names, the value of the  variable ${APP_LANGUAGES} is shown: `en-US.lproj/* es-US.lproj/*` **space separated**. 

Is it clear? 👍

Author: in comments I can drag 'n drop images. In the `.md` I need to host them.

Reviewer 3: To clarify, by default all `lproj` files are excluded aside from the ones in _Include Source File Names_?

Author: @j531 yes, we exclude anything and re add the specified ones in the array.

### Edit a key/value
Search and select the key you'd like to update and change any language you need. The languages you may not have the values will be marked as `Not-verified` and the translators will take care.

## How to add a new Locale
- In Lokalise project, click on the `+` plus button beside other flags. Find the desired language on the list, and then add it.
- In the iOS project, in the file `lokalise` (under the fastlane folder), add the language code (`en_US` for example) in the `langs` parameter for the desired target. `langs: 'es_US,en_US'` (comma separated without space). 

## Pull from lokalise
Check [Lokalise pull guide](https://github.com/Babylonpartners/ios-playbook/blob/master/Cookbook/Technical-Documents/Lokalise.md)

## Commit just your changes
Stage just your changes (additions and editions) on git. Don't stage any additions nor change that you don't recognize. It's really nice to be proactive but not in this case. It might cause unspected/premature changes. Yes, discard others's changes.

## Target specific localizable - Lokalise project and `strings` files

> TBD

Author: [DRAFT]
What is this? 
Why do we need it? 
How to create a key/value that goes to this file? (lokalise key configuration + fastlane output file location)

Author: https://github.com/Babylonpartners/ios-playbook/pull/174/commits/702ba212927ed814a3540f26dab2027aa358486a

## Lokalise: add comments and screenshots when possible.

> TBD

Author: [DRAFT]
What's the value of leaving a comment? 

And a screenshot? Why?
